File: c/dsa/max-depth-binary-tree/max_depth_binary_tree.c
// Max depth of a binary tree (recursive).
/*
        4       depth = 3
      /   \
     2     7
    / \   / \
   1   3 6   9
*/
// Build:  cc -O2 -Wall -Wextra -o depth max_depth_binary_tree.c
// Run:    ./depth

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int val;
    struct Node *left, *right;
} Node;

static Node *new_node(int val, Node *left, Node *right) {
    Node *n = malloc(sizeof(*n));
    n->val = val;
    n->left = left;
    n->right = right;
    return n;
}

static int max_depth(Node *root) {
    if (!root) return 0;
    int l = max_depth(root->left);
    int r = max_depth(root->right);
    return 1 + (l > r ? l : r);
}

static void free_tree(Node *root) {
    if (!root) return;
    free_tree(root->left);
    free_tree(root->right);
    free(root);
}

int main(void) {
    Node *root = new_node(4,
        new_node(2, new_node(1, NULL, NULL), new_node(3, NULL, NULL)),
        new_node(7, new_node(6, NULL, NULL), new_node(9, NULL, NULL)));

    int depth = max_depth(root);
    printf("Tree depth: %d\n", depth);
    assert(depth == 3);
    printf("OK — depth matches expected\n");

    free_tree(root);
    return 0;
}

File: zig/dsa/max-depth-binary-tree/max_depth_binary_tree.zig
// Max depth of a binary tree (recursive).
//
//         4       depth = 3
//       /   \
//      2     7
//     / \   / \
//    1   3 6   9
//
// Build/run:  zig run max_depth_binary_tree.zig

const std = @import("std");

const Node = struct {
    val: i32,
    left: ?*Node = null,
    right: ?*Node = null,
};

fn newNode(alloc: std.mem.Allocator, val: i32, left: ?*Node, right: ?*Node) !*Node {
    const n = try alloc.create(Node);
    n.* = .{ .val = val, .left = left, .right = right };
    return n;
}

fn maxDepth(root: ?*Node) i32 {
    const r = root orelse return 0;
    const l = maxDepth(r.left);
    const rd = maxDepth(r.right);
    return 1 + @max(l, rd);
}

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const a = arena.allocator();

    const root = try newNode(a, 4,
        try newNode(a, 2,
            try newNode(a, 1, null, null),
            try newNode(a, 3, null, null)),
        try newNode(a, 7,
            try newNode(a, 6, null, null),
            try newNode(a, 9, null, null)));

    const depth = maxDepth(root);
    std.debug.print("Tree depth: {d}\n", .{depth});
    std.debug.assert(depth == 3);
    std.debug.print("OK — depth matches expected\n", .{});
}
